core/src/main/java/org/jruby/RubyObject.java

```diff
@@ -152,18 +153,28 @@ public IRubyObject allocate(Ruby runtime, RubyClass klass) {
     public static final ObjectAllocator IVAR_INSPECTING_OBJECT_ALLOCATOR = new ObjectAllocator() {
         @Override
         public IRubyObject allocate(Ruby runtime, RubyClass klass) {
-            Set<String> foundVariables = klass.discoverInstanceVariables();
+            synchronized (klass.getRealClass()) {
+                ObjectAllocator allocator = klass.getAllocator();
 
-            if (Options.DUMP_INSTANCE_VARS.load()) {
-                System.err.println(klass + ";" + foundVariables);
-            }
+                if (allocator == this) {
+                    // proceed, we are the first one to get here
+                    Set<String> foundVariables = klass.discoverInstanceVariables();
+
+                    if (Options.DUMP_INSTANCE_VARS.load()) {
+                        System.err.println(klass + ";" + foundVariables);
+                    }
 
-            ObjectAllocator allocator = RubyObjectSpecializer.specializeForVariables(klass, foundVariables);
+                    allocator = RubyObjectSpecializer.specializeForVariables(klass, foundVariables);
 
-            // invalidate metaclass so new allocator is picked up for specialized .new
-            klass.metaClass.invalidateCacheDescendants();
+                    // invalidate metaclass so new allocator is picked up for specialized .new
+                    klass.metaClass.invalidateCacheDescendants();
 
-            return allocator.allocate(runtime, klass);
+                    return allocator.allocate(runtime, klass);
+                } else {
+                    // someone else replaced allocator while we waited for lock, use it
+                    return allocator.allocate(runtime, klass);
+                }
```

Comment:
Maybe it would affect readability, but would it be possible to get rid of the `else` and always call `return allocator.allocate(runtime, klass);` at the bottom? Not sure if it would have any benefits over just not having duplicated code though (if it is even possible.)

Reply:
Agreed.